### PR TITLE
[SPARK-44918][SQL][PYTHON] Support named arguments in scalar Python/Pandas UDFs

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Supports named arguments in scalar Python/Pandas UDF.

For example:

```py
>>> @udf("int")
... def test_udf(a, b):
...     return a + 10 * b
...
>>> spark.udf.register("test_udf", test_udf)

>>> spark.range(2).select(test_udf(b=col("id") * 10, a=col("id"))).show()
+---------------------------------+
|test_udf(b => (id * 10), a => id)|
+---------------------------------+
|                                0|
|                              101|
+---------------------------------+

>>> spark.sql("SELECT test_udf(b => id * 10, a => id) FROM range(2)").show()
+---------------------------------+
|test_udf(b => (id * 10), a => id)|
+---------------------------------+
|                                0|
|                              101|
+---------------------------------+
```

or:

```py
>>> @pandas_udf("int")
... def test_udf(a, b):
...     return a + 10 * b
...
>>> spark.udf.register("test_udf", test_udf)

>>> spark.range(2).select(test_udf(b=col("id") * 10, a=col("id"))).show()
+---------------------------------+
|test_udf(b => (id * 10), a => id)|
+---------------------------------+
|                                0|
|                              101|
+---------------------------------+

>>> spark.sql("SELECT test_udf(b => id * 10, a => id) FROM range(2)").show()
+---------------------------------+
|test_udf(b => (id * 10), a => id)|
+---------------------------------+
|                                0|
|                              101|
+---------------------------------+
```

### Why are the changes needed?

Now that named arguments support was added (https://github.com/apache/spark/pull/41796, https://github.com/apache/spark/pull/42020).

Scalar Python/Pandas UDFs can support it.

### Does this PR introduce _any_ user-facing change?

Yes, named arguments will be available for scalar Python/Pandas UDFs.

### How was this patch tested?

Added related tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @dtenedor @HyukjinKwon @xinrong-meng

File: python/pyspark/sql/tests/pandas/test_pandas_udf_scalar.py [hunk 1]
```diff
@@ -1467,6 +1467,96 @@ def udf(x):
...
+        def test_udf(a, b):
+            return a + 10 * b
+
+        self.spark.udf.register("test_udf", test_udf)
+
+        for i, df in enumerate(
+            [
+                self.spark.range(2).select(test_udf(a=col("id"), b=col("id") * 10)),
+                self.spark.range(2).select(test_udf(b=col("id") * 10, a=col("id"))),
+                self.spark.sql("SELECT test_udf(a => id, b => id * 10) FROM range(2)"),
```
Comment (dtenedor) [inline python/pyspark/sql/tests/pandas/test_pandas_udf_scalar.py:1481]: can we also have a test case with a positional argument first and then a named argument after? We can have a positive version of this test case where the positional argument maps to the first argument of the UDF and the named argument refers to the second, and another negative version where the named argument uses the same name as the first (positional) argument? Same for `test_udf.py`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDFRunner.scala [hunk 2]
```diff
@@ -146,4 +157,30 @@ object PythonUDFRunner {
       }
     }
   }
+
+  def writeUDFs(
+    dataOut: DataOutputStream,
```
Comment (dtenedor) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonUDFRunner.scala:162]: please indent each of these function arguments by 2 more spaces (for a total of +4 spaces each) per the style guide?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala [hunk 3]
```diff
@@ -50,6 +50,19 @@ case class UserDefinedPythonFunction(
     udfDeterministic: Boolean) {
 
   def builder(e: Seq[Expression]): Expression = {
+    if (pythonEvalType == PythonEvalType.SQL_BATCHED_UDF
+        || pythonEvalType ==PythonEvalType.SQL_ARROW_BATCHED_UDF
+        || pythonEvalType == PythonEvalType.SQL_SCALAR_PANDAS_UDF) {
+      /*
+       * Check if the named arguments:
+       * - don't have duplicated names
+       * - don't contain positional arguments
```
Comment (dtenedor) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:59]:
```suggestion
       * - don't contain positional arguments after named arguments
       * - all map to valid argument names from the function declaration
```

**Review by dtenedor — APPROVED** [review]
Generally LGTM. I left one suggestion for a testing idea. Thanks a lot for adding this feature!!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala (hunk 3 above)
Comment (ueshin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:59]: The third item is not right for Python UDF/UDTF. Currently it relies on Python.

**Review by HyukjinKwon — APPROVED** [review]

**Review by allisonwang-db — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**